crates/ruff/src/commands/format.rs

```diff
@@ -628,6 +670,155 @@ impl<'a> FormatResults<'a> {
+            // individually, otherwise we could use `formatted.source_code().count_lines(...)`
+            // in both cases.
+            let (fix, line_count) = if let SourceKind::IpyNotebook(formatted) = formatted
+                && let SourceKind::IpyNotebook(unformatted) = unformatted
+            {
+                notebook_index.insert(path.to_string(), unformatted.index().clone());
+
+                let mut edits = formatted
+                    .cell_offsets()
+                    .ranges()
+                    .zip(unformatted.cell_offsets().ranges())
+                    .filter_map(|(formatted_range, unformatted_range)| {
+                        // Filter out cells that weren't modified. We use `intersect` instead of
+                        // `contains_range` because the full modified range might start or end in
+                        // the middle of a cell:
+                        //
+                        // ```
+                        // | cell 1 | cell 2 | cell 3 |
+                        //     |----------------| modified range
+                        // ```
+                        //
+                        // The intersection will be `Some` for all three cells in this case.
+                        if modified_range
+                            .unformatted
+                            .intersect(unformatted_range)
+                            .is_some()
+                        {
+                            let formatted = &formatted.source_code()[formatted_range];
+                            let edit = if formatted.is_empty() {
+                                Edit::range_deletion(unformatted_range)
+                            } else {
+                                Edit::range_replacement(formatted.to_string(), unformatted_range)
+                            };
+                            Some(edit)
+                        } else {
+                            None
+                        }
+                    });
+
+                let fix = Fix::safe_edits(
+                    edits
+                        .next()
+                        .expect("Formatted files must have at least one edit"),
+                    edits,
+                );
+                let source = formatted.source_code();
+                let line_count = formatted
+                    .cell_offsets()
+                    .ranges()
+                    .filter_map(|range| {
+                        if modified_range.formatted.contains_range(range) {
+                            Some(source.count_lines(range))
+                        } else {
+                            None
+                        }
+                    })
+                    .max()
+                    .unwrap_or_default();
+                (fix, line_count)
+            } else {
+                let formatted_code = &formatted.source_code()[modified_range.formatted];
+                let edit = if formatted_code.is_empty() {
```

Comment:
It shouldn't matter so I think it's fine leaving as is but there's the third case where `modified_range.unformatted` is empty (e.g. when adding blank lines between two classes), in which case its an insertion. We could add a `Edit::from_text_and_range(new_text, range)` (with a better name) that does this dance.